README.md
NURLY
=====

Nurly is a Nagios Event Broker module that allows service checks to be distributed via libcurl to any compliant http server.  The backend server receiving service check requests is responsible for executing the requested plugin, returning a properly mapped HTTP response code and the output.


Downloading
-----------

- Latest module source: [nurly-module-1.2.5.tar.gz](http://d2g50xfw6d567v.cloudfront.net/mk23/nurly/nurly-module-1.2.5.tar.gz)
- Latest server source: [nurly-server-1.2.5.tar.gz](http://d2g50xfw6d567v.cloudfront.net/mk23/nurly/nurly-server-1.2.5.tar.gz)


Compiling
---------

The build process is through the use of standard autotools command.  If the source is freshly cloned from the repo, first run `autoreconf -i` and then build the module:

    $ ./configure
    $ make
    $ make install

The module will be installed as `/usr/lib/nurly_neb.so`.


Configuration
-------------

### nagios

Nagios needs to have the Execution Broker Module configured and Broker Options set to allow Program State and Service Checks.  Start by editing the main nagios config file (`/etc/nagios3/nagios.cfg`). The parameter to the module is the location of nurly's config file.

    # EVENT BROKER MODULE(S)
    # This directive is used to specify an event broker module that should
    # by loaded by Nagios at startup.  Use multiple directives if you want
    # to load more than one module.  Arguments that should be passed to
    # the module at startup are seperated from the module path by a space.
    broker_module=/usr/lib/nurly_neb.so /etc/nagios3/nurly.cfg

    # EVENT BROKER OPTIONS
    # Controls what (if any) data gets sent to the event broker.
    # Values:  0      = Broker nothing
    #         -1      = Broker everything
    #         <other> = See documentation
    event_broker_options=-1

### nurly

The nurly configuration file is rather self-explanatory.

    ## url for regular service checks
    checks_url = http://localhost:2323/check-command/

    ## url for backend health check
    health_url = http://localhost:2323/server-version

    ## health check interval in seconds
    health_interval = 5

    ## number of nagios worker threads
    worker_threads = 20

    ## seconds before curl gives up and check results in UNKNOWN
    http_timeout = 10

    ## do not distribute checks for these hosts
    skip_host = localhost
    skip_host = special[0-9].*\.com

    ## do not distribute checks for these services
    skip_service = nagios_config
    skip_service = ^SSH$

Operation
---------

The nurly module brokers service checks from Nagios by appending the expanded command-line that Nagios would have otherwise run locally to the url configured and performing the request via libcurl.  This happens in a separate thread, allowing the rest of Nagios to continue operating without waiting for a reply.  The server is expected to perform the requested check and return the output of the check as the body with a specific status code indicating check result.  Nurly recognizes the following HTTP status codes and converts them to proper Nagios result codes:

    HTTP Response    Nagios Result
    200 OK           0 Success/OK
    221 Warning      1 Warning
    222 Critical     2 Critical
    223 Unknown      3 Unknown


Server
------

An example nurly server exists in the `server` subdirectory.  Its feature set includes:

- multiprocess (multi-core) operation, with customizable number of workers
- customizable whitelist of allowed plugin directories
- customizable whitelist of allowed hosts
- embeddable python plugin execution by importing and calling `main()` saving `fork()/exec()` per check

The Nurly server is fully configurable via command-line parameters:

    $ server/nurly.py --help
    usage: nurly.py [-h] [-a SERVER_ADDR] [-p SERVER_PORT] [-n NUM_WORKERS]
                    [-i ALLOWED_IPS [ALLOWED_IPS ...]]
                    [-l PLUGIN_PATH [PLUGIN_PATH ...]] [-t MOD_TIMEOUT]

    nagios service check handler

    optional arguments:
      -h, --help            show this help message and exit
      -a SERVER_ADDR, --server-addr SERVER_ADDR
                            local listening interface (default: )
      -p SERVER_PORT, --server-port SERVER_PORT
                            local listening port (default: 1123)
      -n NUM_WORKERS, --num-workers NUM_WORKERS
                            number of worker processes (default: 2)
      -i ALLOWED_IPS [ALLOWED_IPS ...], --allowed-ips ALLOWED_IPS [ALLOWED_IPS ...]
                            allowed host whitelist, permits all when empty
                            (default: [])
      -l PLUGIN_PATH [PLUGIN_PATH ...], --plugin-path PLUGIN_PATH [PLUGIN_PATH ...]
                            path to plugin directory (default:
                            ['/usr/lib/nagios/plugins'])
      -t MOD_TIMEOUT, --mod-timeout MOD_TIMEOUT
                            module execution timeout (default: 10)


License
-------

TBD


debian/changelog
nurly (1.2.5) precise; urgency=low

  * Tagging 1.2.5
  * [b1b96c6] Issue #17: Fix http_timeout config value processing.

 -- Max Kalika <max@topsy.com>  Fri, 10 May 2013 17:45:25 -0700

nurly (1.2.4) precise; urgency=low

  * Tagging 1.2.4
  * [a92612d] Issue #16: Add an example nurly-compatible php script.
  * [75bb2d3] Issue #15: Fix forking in compatible plugins.
  * [b4fcb03] Consolidate git ignores.
  * [488c5ee] Issue #14: Fix SIGSEGV by joining the health and worker
    threads during module deinit.

 -- Max Kalika <max@topsy.com>  Thu, 09 May 2013 04:15:57 -0700

nurly (1.2.3) precise; urgency=low

  * Tagging 1.2.3
  * [a90e5e5] Issue #8: Standardize logging to nagios messages types and
    debug levels.
  * [36b6c5f] Issue #9: add flag to turn off access logging.
  * [1bc80f1] Issue #11: Make the curl http timeout configurable.
  * [6c2fe03] Issue #12: Fix double decoding of request paths.
  * [13bb0d5] Issue #13: Suspend SIGCHLD handler while running
    check_command because it may interfere with the subprocess module.

 -- Max Kalika <max@topsy.com>  Fri, 26 Apr 2013 14:13:29 -0700

nurly (1.2.2) precise; urgency=low

  * Tagging 1.2.2
  * [8834375] Fix python version in debian control.
  * [800e882] Fix server usage in the doc too.

 -- Max Kalika <max@topsy.com>  Thu, 25 Apr 2013 09:21:14 -0700

nurly (1.2.1) precise; urgency=low

  * Tagging 1.2.1
  * [3f6ceae] Fix help text for --allowed-ips in httpd.py.

 -- Max Kalika <max@topsy.com>  Wed, 24 Apr 2013 23:39:45 -0700

nurly (1.2.0) precise; urgency=low

  * Tagging 1.2.0
  * [60a92e4] Update links in docs when cutting a release.
  * [18b8fd5] Format the links a bit better in the docs.
  * [9425fbb] Update docs to the latest nurly server code.
  * [62fb461] Add links to downloadable source in the docs.
  * [a127434] Issue #7: Update setup.py to new package layout.
  * [c94b3cd] Move version directly into httpd.py.
  * [71d8008] Issue #7: Purge the old wsgi-based server code.
  * [1826765] Issue #7: Convert nurly server to an httpd module.
  * [35f0d84] Issue #7: Make the extra settings attribute a named tuple
    for easier access and immutability.
  * [2b78fc5] Issue #7: Simplify server creation with command-line args
    auto expansion.
  * [a544389] Issue #7: Propegate arbitrary external configuration to
    the handler.
  * [bb0a3cd] Issue #7: Better module name.
  * [f9b64a1] Clean up children on SIGTERM.
  * [1967157] Simplify construction and server start.
  * [d1288ae] Add ip whitelist support.
  * [297ecd7] Simplify http error exceptions.
  * [4b7ac47] Add start of command-line args for the library to be used
    as a parent parser for importers.
  * [a77b032] Add server version request handler.
  * [140c303] Get rid of test request handlers.
  * [c5b2715] Add scoreboard support to the http library.
  * [e5d5378] Simplify decorator for plain path request handlers by
    using the function name.
  * [5c97c69] Issue #7: Start on a simplified server library.

 -- Max Kalika <max@topsy.com>  Wed, 24 Apr 2013 23:10:34 -0700

nurly (1.1.0) precise; urgency=low

  * Tagging 1.1.0
  * [7adc598] Issue #6: Add support for python plugins that don't
    implement main().
  * [ef16d7c] Generate the server tarball alongside module tarball when
    calling 'make dist'
  * [1b6a0d6] Add docs for skip_host/service config items.
  * [53236e7] Issue #5: Return error on memory failures when adding to
    the queue.
  * [41ec380] Autoconf check for regex header and functions.
  * [c7a95c8] Log consistency.
  * [c5de21e] Issue #4: Hook up skip_host/service configuration item to
    avoid check distribution if requested by operator.
  * [f26cfb6] Organize functions better.
  * [61f51fd] Ignore case in skip_host/service patterns.
  * [932031f] Avoid infinite loop.
  * [c80dcf2] Add nurly_queue_has() callback to perform regex match.
  * [0be1269] Add config processor for skip_service setting.
  * [a40d6b2] Ignore the rest of the autoconf vomit.
  * [00f8504] New queue function: nurly_queue_has(): searches the queue
    via callback for every item.
  * [3e28398] Issue #4: Parse new skip_host and skip_service settings,
    compile as regex, and add to internal lists.
  * [3c36b1c] Issue #4: Add skip_hosts and skip_services config members.
  * [309dfeb] No need to cast return of strdup() as char*.
  * [98b7590] Whitespace.
  * [69e91df] More documentation.
  * [7d063a9] Pass extra flags (i.e. --minor, --major, etc.) to
    release.py.
  * [0b7cf61] Add initial readme file.
  * [f8bd430] Change default prefix to /usr.
  * [ba97c1e] Update email and url.
  * [25e6210] Issue #2: Finalize health check system.
  * [d2965cb] Issue #2: Stub out health check thread.
  * [95b0046] Fix log grammar.
  * [517cfbd] Reorganize config management a bit.

 -- Max Kalika <max@topsy.com>  Tue, 23 Apr 2013 08:27:53 -0700

nurly (1.0.1) precise; urgency=low

  * Tagging 1.0.1
  * [60be578] Finalize release.py integration.
  * [617bc29] Issue #1: add release management via release.py.
  * [a667fd9] Static version file for the server.
  * [c87484b] autoconf will no longer handle server's version.

 -- Max Kalika <max@topsy.com>  Tue, 12 Mar 2013 13:34:42 -0700

nurly (1.0.0) precise; urgency=low

  * Initial version.

 -- Max Kalika <max@mklinbox.vm-zone>  Mon, 21 Jan 2013 23:23:39 -0800


server/http.py
#!/usr/bin/env python2.7

import BaseHTTPServer

import argparse
import collections
import multiprocessing
import os
import re
import select
import signal
import socket
import sys
import time
import traceback
import urllib
import urlparse

from version import VERSION

STATUS_IDLE, STATUS_BUSY, STATUS_STOP = 'IBS'

class HTTPError(Exception):
    code = 500
    def __init__(self, exception=None):
        self.line, self.body = Handler.responses[self.code]
        if exception is not None:
            self.body = ''.join(traceback.format_exception(*exception))

class HTTPForbiddenError(HTTPError):
    code = 403

class HTTPNotFoundError(HTTPError):
    code = 404

class Response():
    def __init__(self, **kwargs):
        self.__dict__.update({
            'code': 200,
            'line': 'OK',
            'head': collections.OrderedDict(),
            'body': '',
        })

        for key, val in kwargs.items():
            if key == 'head':
                self.head.update(kwargs['head'])
            else:
                setattr(self, key, val)

    def __setattr__(self, key, val):
        if key.startswith('_') or key == 'head' or key not in self.__dict__:
            raise AttributeError('setting attribute %s is not supported' % key)
        if key == 'code' and type(val) != int:
            raise AttributeError('setting attribute %s requires an integer' % key)

        self.__dict__[key] = val

class Worker(multiprocessing.Process):
    def __init__(self, server, worker_pipe, parent_pipe, unused_pipes):
        self.channel = parent_pipe
        self.server = server
        self.fileno = self.channel.fileno

        unused_pipes.append(parent_pipe)

        multiprocessing.Process.__init__(self, target=self.start_server, args=(worker_pipe, unused_pipes))
        self.start()

    def reap_children(self, *args):
        os.waitpid(-1, os.WNOHANG)

    def send_channel(self, func, args=[], recv=False):
        if type(args) not in (list, tuple):
            args = [args]
        self.channel.send( (func, args) )

        if recv:
            return self.channel.recv()

    def start_server(self, worker_pipe, unused_pipes):
        signal.signal(signal.SIGTERM, signal.SIG_DFL)
        signal.signal(signal.SIGCHLD, self.reap_children)

        self.channel = worker_pipe
        self.fileno = self.channel.fileno

        try:
            for pipe in unused_pipes:
                pipe.close()

            self.server.worker = self
            self.server.serve_forever(poll_interval=None)
        except KeyboardInterrupt:
            return

class Server(BaseHTTPServer.HTTPServer):
    allow_reuse_address = True

    def __init__(self, port, addr='', workers=1, allowed=None, version=None):
        BaseHTTPServer.HTTPServer.__init__(self, (addr, port), Handler)

        self.unused_pipes = []
        self.worker_state = {}
        self.ip_whitelist = set(socket.gethostbyname(i) for i in allowed)

        Handler.server_version += ' %s/%s' % (os.path.basename(sys.modules[__name__].__file__), VERSION)
        if version is not None:
            Handler.server_version += ' %s' % version

        self.create_server(workers)

    def server_bind(self):
        self.socket.setsockopt(socket.SOL_TCP, socket.TCP_DEFER_ACCEPT, True)
        self.socket.setsockopt(socket.SOL_TCP, socket.TCP_QUICKACK,     True)
        BaseHTTPServer.HTTPServer.server_bind(self)

    def create_server(self, workers):
        signal.signal(signal.SIGTERM, self.clean_workers)

        self.worker_state.update(self.create_worker() for i in xrange(workers))
        self.handle_events()

    def create_worker(self):
        parent_pipe, worker_pipe = multiprocessing.Pipe(duplex=True)

        worker_proc = Worker(self, worker_pipe, parent_pipe, self.unused_pipes)
        worker_proc.status = STATUS_IDLE
        worker_pipe.close()

        return worker_proc.pid, worker_proc

    def clean_workers(self, *args):
        for pid in self.worker_state.keys():
            os.kill(pid, signal.SIGTERM)

        sys.exit(0)

    def ch_set_status(self, worker, status):
        worker.status = status

    def ch_get_status(self, worker):
        worker.channel.send(list(i.status for i in self.worker_state.values()))

    def handle_events(self):
        while True:
            for worker in select.select(self.worker_state.values(), [], [])[0]:
                try:
                    func, args = worker.channel.recv()
                    getattr(self, 'ch_%s' % func, lambda *a: True)(worker, *args)
                except EOFError:
                    worker.join()
                    del self.worker_state[worker.pid]

                    pid, proc = self.create_worker()
                    self.worker_state[pid] = proc
                except Exception:
                    traceback.print_exc()


class Handler(BaseHTTPServer.BaseHTTPRequestHandler):
    protocol_version = 'HTTP/1.1'

    request_handlers = {
        'GET': collections.OrderedDict()
    }

    default_headers = {
        'Connection': 'close',
    }

    def route_request(self, method):
        self.start_time  = time.time() * 1000

        self.server.worker.send_channel('set_status', STATUS_BUSY)

        self.nurly_url   = urlparse.urlparse(self.path)
        self.nurly_path  = urllib.unquote(self.nurly_url.path)
        self.nurly_query = urlparse.parse_qs(self.nurly_url.query)

        response = Response(head=self.default_headers)

        try:
            if self.server.ip_whitelist and self.client_address[0] not in self.server.ip_whitelist:
                raise HTTPForbiddenError

            for patt, func in self.request_handlers[method].items():
                find = patt.match(urllib.unquote(self.nurly_path))
                if find:
                    response = Response(head=self.default_headers)
                    func(self, response, *(find.groups()))
                    break
            else:
                raise HTTPNotFoundError

        except Exception as e:
            if not isinstance(e, HTTPError):
                e = HTTPError(sys.exc_info())
            response.code = e.code
            response.line = e.line
            response.body = e.body

        self.send_response(response.code, response.line)
        for key, val in response.head.items():
            if key.lower() != 'content-length':
                self.send_header(key, val)
        self.send_header('Content-length', len(response.body))
        self.send_header('X-Handler-Time', '%.02fms' % (time.time() * 1000 - self.start_time))
        self.send_header('X-Handler-Pid', os.getpid())
        self.end_headers()
        self.wfile.write(response.body)

        self.server.worker.send_channel('set_status', STATUS_IDLE)

    def do_GET(self):
        self.route_request('GET')

def GET(path=None):
    def wrapper(func):
        patt = r'/%s(?:/|$)' % (path.strip('/') if path is not None else func.func_name.replace('_', '-'))
        Handler.request_handlers['GET'][re.compile(patt)] = func
    return wrapper


@GET()
def server_status(req, res):
    status = req.server.worker.send_channel('get_status', recv=True)
    res.body = 'Busy Workers: %d\nIdle Workers: %d\nScoreboard:   %s\n' % (
        status.count(STATUS_BUSY), status.count(STATUS_IDLE), ''.join(status)
    )

@GET()
def server_version(req, res):
    res.body = req.server_version + '\r\n'

parser = argparse.ArgumentParser(description='python http server library', formatter_class=argparse.ArgumentDefaultsHelpFormatter)
parser.add_argument('-p', '--server-port', default=1123, type=int,
                    help='local listening port')
parser.add_argument('-n', '--num-workers', default=multiprocessing.cpu_count(), type=int,
                    help='number of worker processes')
parser.add_argument('-a', '--allowed-ips', default=[], nargs='+',
                    help='number of worker processes')

if __name__ == '__main__':
    args = parser.parse_args()

    Server(args.server_port, allowed=args.allowed_ips, workers=args.num_workers)
